File: 5.longest-palindromic-substring.ts
/*
 * @lc app=leetcode id=5 lang=typescript
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
function longestPalindrome(s: string): string {
    let longestPalindrome = s[0] ?? '';

    const checkPalindrome = (leftCursor, rightCursor) => {
        for (let j = 0; j < s.length; j++) {
            leftCursor -= 1;
            rightCursor += 1;
            if (leftCursor < 0) break;
            if (rightCursor >= s.length) break;
            if (s[leftCursor] !== s[rightCursor]) break;
        }
        if (longestPalindrome.length < (rightCursor - leftCursor - 1)) {
            longestPalindrome = s.slice(leftCursor + 1, rightCursor);
        }
    }

    for (let i = 0; i < s.length - Math.ceil(longestPalindrome.length / 2); i++) {
        checkPalindrome(i, i);
        if (s[i] === s[i + 1]) checkPalindrome(i, i + 1);
    }

    return longestPalindrome;
};
// @lc code=end

File: 2024/5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#

# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        def checkPalindrome(left, right):
            if s[left] != s[right]:
                return
            while left > 0 and right + 1 < len(s) and s[left - 1] == s[right + 1]:
                left -= 1
                right += 1
            if right - left + 1 > len(self.result):
                self.result = s[left:right+1]

        self.result = ''
        for i in range(len(s)):
            checkPalindrome(i, i)
            if i + 1 < len(s):
                checkPalindrome(i, i+1)

        return self.result
# @lc code=end

